ENH: Adding support of GridSearchCV for mne.decoding.LinearModel

iamsc opened the pull request:
#### Reference issue
Fixes #7027 .

#### What does this implement/fix?
This PR adds the support of `GridSearchCV` for `LinearModel`. I also ran the decoding/tests/test_base.py and got the 3 passed.

agramfort commented:
@iamsc can you please add a new test in  https://github.com/mne-tools/mne-python/blob/master/mne/decoding/tests/test_base.py ? thx

iamsc commented:
> @iamsc can you please add a new test in https://github.com/mne-tools/mne-python/blob/master/mne/decoding/tests/test_base.py ? thx

No problem~ I have modified some tests in `test_base.py`.

Because the original `test_get_coef` function in `test_base.py` only focused on the regression estimator tests, I added some new tests which considers some possible types of estimators, likes classification, regression, classification with GridSearchCV, regression with GridSearchCV.

Moreover, in `test_linearmodel` function, I also added the classification/regression with GridSearchCV tests.

agramfort commented:
@iamsc you have some flake8 errors. Run `make pep` or flake8 on your files to see what's wrong

iamsc commented:
> @iamsc you have some flake8 errors. Run `make pep` or flake8 on your files to see what's wrong

Sorry for those errors. I have already fixed those errors on my files using `make pep`.

However, there are still some other errors from `make pep`, likes `mne/tests/test_docstring_parameters.py`. I just ignored this kind of errors because I have not modified those files in this PR.

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -61,13 +61,32 @@ def test_get_coef():
+    lm_gs_classification = LinearModel(
+        GridSearchCV(
+            svm.SVC(), parameters,
+            cv=5, refit=True, iid=False, n_jobs=5
```
for testing try to use something as fast as possible. So cv=2. Also n_jobs should be set to 1 for testing.

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -61,13 +61,32 @@ def test_get_coef():
+    lm_gs_regression = LinearModel(
+        GridSearchCV(
+            svm.SVR(), parameters, cv=5,
+            refit=True, iid=False, n_jobs=5
```
same here

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -181,10 +232,39 @@ def test_linearmodel():
+    clf = LinearModel(
+        GridSearchCV(
+            svm.SVC(), parameters, cv=5,
+            refit=True, iid=False, n_jobs=5
```
same here

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -181,10 +232,39 @@ def test_linearmodel():
+    clf = LinearModel(
+        GridSearchCV(
+            svm.SVR(), parameters, cv=5,
+            refit=True, iid=False, n_jobs=5
```
same

iamsc commented:
> tests should be as fast as possible. Can you report how long it takes to run now compared to before?

If it only sets `cv` as 2, the test duration will decrease 77.17 seconds compared to before.

```
========================================================================== slowest 20 test durations ===========================================================================
180.56s call     mne/decoding/tests/test_base.py::test_get_coef
3.36s setup    mne/decoding/tests/test_base.py::test_get_coef
0.50s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.08s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel

Results (184.80s):
       3 passed
```

If I remove 1000 from `C` list in parameter grids and set `cv`  as 2, the test duration will decrease to 15.92 seconds.

```
========================================================================== slowest 20 test durations ===========================================================================
11.61s call     mne/decoding/tests/test_base.py::test_get_coef
3.60s setup    mne/decoding/tests/test_base.py::test_get_coef
0.19s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.03s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel

Results (15.69s):
       3 passed
```

I will re-commit the second one later.

larsoner commented:
Even 11 sec is quite long for a test. How long is it in `master`? Ideally these changes wouldn't add more than 0.1 sec or so. We run a lot of tests, and a lot that do much more complex processing than this, so we need to keep test times to a minimum.

If you need help speeding it up let me know and I can look at tweaking the test to get the time down

iamsc commented:
> Even 11 sec is quite long for a test. How long is it in `master`? Ideally these changes wouldn't add more than 0.1 sec or so. We run a lot of tests, and a lot that do much more complex processing than this, so we need to keep test times to a minimum.
> 
> If you need help speeding it up let me know and I can look at tweaking the test to get the time down

Here is the time duration results of `master`:
```
========================================================================== slowest 20 test durations ===========================================================================
3.59s setup    mne/decoding/tests/test_base.py::test_get_coef
0.70s call     mne/decoding/tests/test_base.py::test_get_coef
0.23s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.01s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel

Results (4.82s):
       3 passed
```

I also tried to remove 100 from `C` list further, the results are as follows:
```
========================================================================== slowest 20 test durations ===========================================================================
4.07s setup    mne/decoding/tests/test_base.py::test_get_coef
1.32s call     mne/decoding/tests/test_base.py::test_get_coef
0.21s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.02s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef

Results (6.03s):
       3 passed
```
It seems like to close to the time duration in `master`. I come up with some possible directions I can try with, such as reducing the testing sample size, removing some tests for the possible types of estimators, etc. @larsoner do you have other suggestions that I can also try with?

larsoner commented:
I would try those, plus reducing the number of samples and/or features.

If the regressors/estimators you are iterating over are `sklearn` estimators, you probably don't need to do more than one. We can trust them to adhere to their own API requirements, so testing a single one should work.

iamsc commented:
Thanks @larsoner for your suggestions. I have tweaked some tests, including removing some estimators, reducing the parameter space, and reducing the number of samples.

Here is the new time duration:
```
========================================================================== slowest 20 test durations ===========================================================================
3.22s setup    mne/decoding/tests/test_base.py::test_get_coef
0.79s call     mne/decoding/tests/test_base.py::test_get_coef
0.21s call     mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.02s call     mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_linearmodel
0.00s setup    mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_cross_val_multiscore
0.00s teardown mne/decoding/tests/test_base.py::test_linearmodel
0.00s teardown mne/decoding/tests/test_base.py::test_get_coef

Results (4.43s):
       3 passed
```

iamsc commented:
OK~ No problem.

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -114,16 +133,43 @@ def inverse_transform(self, X):
+        # according to the type of estimator.
+        if is_classifier(clf):
+            n, n_features = 1000, 3
+            X = np.random.rand(n, n_features)
```
Please use a random state to this test has no effect on global state

iamsc replied:
Sorry for the late replay.
Does this mean that it would be better to use the `np.random.seed(123)` or `np.random.RandomState(123)` before the `np.random.rand()` function?

larsoner replied:
Typically we try to do:
```
def test_whatever():
    rng = np.random.RandomState(0)
    ...
       X = rng.rand(n, n_features)
```

iamsc replied:
Oh! I see~ Let me modify it.

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -181,10 +228,39 @@ def test_linearmodel():
+    clf.fit(X, y)
+    assert_equal(clf.filters_.shape, (n_features,))
+    assert_equal(clf.patterns_.shape, (n_features,))
+    pytest.raises(ValueError, clf.fit, np.random.rand(n, n_features, 99), y)
```
Same here

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -181,10 +228,39 @@ def test_linearmodel():
+
+    # check continuous target fit in standard linear model with GridSearchCV
+    n_targets = 1
+    Y = np.random.rand(n, n_targets)
```
Idem

agramfort commented on mne/decoding/tests/test_base.py:
```diff
@@ -181,10 +228,39 @@ def test_linearmodel():
     assert_equal(clf.patterns_.shape, (n_features,))
     pytest.raises(ValueError, clf.fit, np.random.rand(n, n_features, 99), y)
```
Its now recommended to use with block to catch exceptions

agramfort commented:
Thx @iamsc !

iamsc commented:
I’m happy to contribute to MNE and learn from you folks. Thanks @agramfort and @larsoner .

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

agramfort commented:
looking forward to your next PRs :)

thanks for contributing

>
